lib/iterator.js

```diff
@@ -0,0 +1,317 @@
+      if (this.predicate.call(this.thisArg, value)) {
+        return { done: false, value };
+      }
+    }
+    return { done: true, value: undefined };
+  }
+}
+
+class FlatIterator extends Iterator {
+  constructor(base, depth) {
+    super(base);
+    this.depth = 0;
+    this.stack = new Array(depth + 1);
+    this.stack[0] = base;
+  }
+
+  next() {
+    while (this.depth >= 0) {
+      const top = this.stack[this.depth];
+      const next = top.next();
+
+      if (next.done) {
+        this.stack[this.depth] = undefined;
+        this.depth--;
+        continue;
+      }
+
+      if (this.depth === this.stack.length - 1) {
+        return next;
+      }
+
+      if (next.value[Symbol.iterator]) {
+        this.stack[++this.depth] = next.value[Symbol.iterator]();
+        continue;
+      }
+
+      if (!next.value.next) {
+        return next;
+      }
+
+      this.stack[++this.depth] = next.value[Symbol.iterator]();
+    }
+
+    return { done: true, value: undefined };
+  }
+}
+
+class FlatMapIterator extends Iterator {
+  constructor(base, mapper, thisArg) {
+    super(base);
+    this.mapper = mapper;
+    this.thisArg = thisArg;
+    this.currentIterator = undefined;
+  }
+
+  next() {
+    if (!this.currentIterator) {
+      const next = this.base.next();
+      if (next.done) {
+        return next;
+      }
+
+      const value = this.mapper.call(this.thisArg, next.value);
+      if (!value[Symbol.iterator] && !value.next) {
+        return { done: false, value };
+      }
+
+      this.currentIterator = new Iterator(value);
+    }
+
+    const next = this.currentIterator.next();
+
+    if (next.done) {
+      this.currentIterator = undefined;
+      return this.next();
+    }
+    return next;
+  }
+}
+
+class TakeIterator extends Iterator {
+  constructor(base, amount) {
+    super(base);
+    this.amount = amount;
+    this.iterated = 0;
+  }
+
+  next() {
+    if (this.iterated++ < this.amount) {
+      return this.base.next();
+    }
+    return { done: true, value: undefined };
+  }
+}
+
+class ZipIterator extends Iterator {
+  constructor(base, iterators) {
+    super(base);
+    this.iterators = iterators.map(toIterator);
+  }
+
+  next() {
+    const result = [];
+
+    const next = this.base.next();
+    if (next.done) {
+      return next;
+    }
+    result.push(next.value);
+
+    for (const iterator of this.iterators) {
+      const next = iterator.next();
+      if (next.done) {
+        return next;
+      }
+      result.push(next.value);
+    }
+    return { done: false, value: result };
+  }
+}
+
+class JoinIterator extends Iterator {
+  constructor(base, iterators) {
+    super(base);
+    this.currentIterator = base;
+    this.iterators = iterators.map(toIterator)[Symbol.iterator]();
+  }
+
+  next() {
+    const next = this.currentIterator.next();
+    if (!next.done) {
+      return next;
+    }
+    const { done, value: iterator } = this.iterators.next();
+    if (done) {
+      return next;
```

Comment:
It's a bit counterintuitive and looks like a typo. Maybe add a comment that `next` is always `{ done: true, value: undefined }` at this point?